File: snake_game/entities.py
"""The Snake entity: movement, growth, and AI planning."""
from .ai import bfs_direction, survival_direction


class Snake:
    """A single snake body with movement and self/obstacle awareness."""

    def __init__(self, body, direction, color, is_ai=False):
        self.body = list(body)          # head is body[0]
        self.direction = direction
        self.pending = direction
        self.color = color
        self.is_ai = is_ai
        self.alive = True
        self.grow_pending = 0
        self.score = 0
        self.cycle = None               # set for Hamiltonian fill AI

    @property
    def head(self):
        return self.body[0]

    def set_direction(self, direction):
        # Disallow reversing directly onto the neck.
        if (direction[0] == -self.direction[0]
                and direction[1] == -self.direction[1]):
            return
        self.pending = direction

    def plan_ai(self, food, blocked):
        """Choose a direction using BFS, falling back to survival."""
        direction = bfs_direction(self.head, food, blocked)
        if direction is None:
            direction = survival_direction(self.head, blocked, self.direction)
        if direction is not None:
            self.set_direction(direction)

    def plan_cycle(self):
        """Follow the Hamiltonian cycle (authoritative; bypasses the guard)."""
        hx, hy = self.head
        nx, ny = self.cycle[self.head]
        self.pending = (nx - hx, ny - hy)

    def step(self):
        self.direction = self.pending
        x, y = self.head
        new_head = (x + self.direction[0], y + self.direction[1])
        self.body.insert(0, new_head)
        if self.grow_pending > 0:
            self.grow_pending -= 1
        else:
            self.body.pop()

    def grow(self, amount=1):
        self.grow_pending += amount

File: js/entities.js
/* The Snake entity: movement, growth, and AI hooks. */
import { key, bfsDirection, survivalDirection } from "./ai.js";

export class Snake {
  constructor(body, dir, color, isAi = false) {
    this.body = body.map((c) => [c[0], c[1]]); // head at index 0
    this.dir = dir;
    this.pending = dir;
    this.color = color;
    this.isAi = isAi;
    this.alive = true;
    this.growPending = 0;
    this.score = 0;
    this.cycle = null;        // set for Hamiltonian fill AI
    this.brain = "bfs";       // AI planner: bfs/astar/greedy/anneal/drift/minimax/survive
  }
  get head() { return this.body[0]; }
  setDirection(dir) {
    if (dir[0] === -this.dir[0] && dir[1] === -this.dir[1]) return;
    this.pending = dir;
  }
  planAi(food, blocked) {
    let dir = food ? bfsDirection(this.head[0], this.head[1], food[0], food[1], blocked) : null;
    if (!dir) dir = survivalDirection(this.head[0], this.head[1], blocked, this.dir);
    if (dir) this.setDirection(dir);
  }
  planCycle() {
    const [hx, hy] = this.head;
    const nxt = this.cycle.get(key(hx, hy));
    this.pending = [nxt[0] - hx, nxt[1] - hy];
  }
  step() {
    this.dir = this.pending;
    const [hx, hy] = this.head;
    this.body.unshift([hx + this.dir[0], hy + this.dir[1]]);
    if (this.growPending > 0) this.growPending -= 1;
    else this.body.pop();
  }
  grow(n = 1) { this.growPending += n; }
  has(x, y, fromIndex = 0) {
    for (let i = fromIndex; i < this.body.length; i++) {
      if (this.body[i][0] === x && this.body[i][1] === y) return true;
    }
    return false;
  }
}
